coalib/bearlib/abstractions/ExternalBearWrap.py

```diff
@@ -0,0 +1,134 @@
+            the executable
+            """
+
+            raise NotImplementedError
+
+        @classmethod
+        def get_executable(cls):
+            """
+            Returns the executable of this class.
+
+            :return: The executable name.
+            """
+            return options["executable"]
+
+        @staticmethod
+        def normalize_desc(val):
+            if val[0] != "":
+                return val
+            desc = "No description given."
+            if len(val) == 2:
+                return (desc, val[1])
+            else:
+                desc += " Defaults to " + str(val[2])
+                return (desc, val[1], val[2])
+
+        @classmethod
+        def get_non_optional_params(cls):
+            dict = {}
+            for key, value in options['settings'].items():
+                if len(value) == 2:
+                    dict[key] = cls.normalize_desc(value)
+            return OrderedDict(dict)
+
+        @classmethod
+        def get_optional_params(cls):
+            dict = {}
+            for key, value in options['settings'].items():
+                if len(value) == 3:
+                    dict[key] = cls.normalize_desc(value)
+            return OrderedDict(dict)
+
+        @classmethod
+        def get_metadata(cls):
+            metadata = FunctionMetadata(
+                'run',
+                optional_params=cls.get_optional_params(),
+                non_optional_params=cls.get_non_optional_params())
+            return metadata
+
+        def parse_output(self, out, filename):
+            try:
+                output = json.loads(out)
+            except ValueError:
+                raise ValueError
+
+            for result in output:
+                yield Result.from_values(
+                    origin=self,
+                    message=result['message'],
+                    file=filename,
+                    line=result['line'])
+
+        def run(self, filename, file, **kwargs):
+            json_string = json.dumps({
+                                    'filename': filename,
+                                    'file': file
+                                    })
+
+            args = self.create_arguments()
+            try:
+                args = tuple(args)
+            except TypeError:
+                self.err("The given arguments "
+                         "{!r} are not iterable.".format(args))
+                return
+
+            shell_command = (self.get_executable(),) + args
+            out, err = run_shell_command(shell_command, json_string)
+
+            for result in self.parse_output(out, filename):
+                yield result
+
+    result_klass = type(klass.__name__, (klass, ExternalBearWrapBase), {})
+    result_klass.__doc__ = klass.__doc__ if klass.__doc__ else ""
+    return result_klass
+
+
+@enforce_signature
+def external_bear_wrap(executable: str,
+                       **options):
```

Comment:
I would put the signature into a single line, makes it easier to read (at least for me :D)